src/compiler/binder.ts

```diff
@@ -329,12 +351,281 @@ namespace ts {
+                currentReachabilityState = ifFalseState;
+                bind(n.elseStatement);
+                currentReachabilityState = or(currentReachabilityState, preElseState);
+            }
+            else {
+                currentReachabilityState = or(currentReachabilityState, ifFalseState);
+            }
+
+            return true;
+        }
+
+        function bindReturnOrThrow(n: ReturnStatement | ThrowStatement): boolean {
+            // bind expression (don't affect reachability)
+            bind(n.expression);
+            if (n.kind === SyntaxKind.ReturnStatement) {
+                hasExplicitReturn = true;
+            }
+            currentReachabilityState = Reachability.Unreachable;
+
+            return true;
+        }
+
+        function bindBreakOrContinueStatement(n: BreakOrContinueStatement): boolean {
+            // call bind on label (don't affect reachability)
+            bind(n.label);
+            if (n.kind === SyntaxKind.BreakStatement) {
+                jumpToLabel(n.label, currentReachabilityState);
+            }
+            else {
+                jumpToLabel(n.label, Reachability.Unreachable);  // touch label so it will be marked a used
+            }
+            currentReachabilityState = Reachability.Unreachable;
+
+            return true;
+        } 
+
+        function bindTryStatement(n: TryStatement): boolean {
+            // catch\finally blocks has the same reachability as try block
+            const preTryState = currentReachabilityState;
+            bind(n.tryBlock);
+            const postTryState = currentReachabilityState;
+
+            currentReachabilityState = preTryState;
+            bind(n.catchClause);
+            const postCatchState = currentReachabilityState;
+
+            currentReachabilityState = preTryState;
+            bind(n.finallyBlock);
+
+            // post catch/finally state is reachable if
+            // - post try state is reachable - control flow can fall out of try block
+            // - post catch state is reachable - control flow can fall out of catch block
+            currentReachabilityState = or(postTryState, postCatchState);
+
+            return true;
+        }
+
+        function bindSwitchStatement(n: SwitchStatement): boolean {
+            const preSwitchState = currentReachabilityState;
+            const postSwitchLabel = pushImplicitLabel();
+
+            // bind expression (don't affect reachability)
+            bind(n.expression);
+
+            bind(n.caseBlock);
+            
+            const hasDefault = forEach(n.caseBlock.clauses, c => c.kind === SyntaxKind.DefaultClause);
+
+            // post switch state is unreachable if switch is exaustive (has a default case ) and does not have fallthrough from the last case
+            const postSwitchState = hasDefault && currentReachabilityState !== Reachability.Reachable ? Reachability.Unreachable : preSwitchState;

```

Comment:
I don't understand reachability of switch-statement. Why being exhaustive prevent post switch to be reachable?